oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexCleaner.java

```diff
@@ -0,0 +1,130 @@
+     * @param thresholdInSeconds time in seconds before which a dangling remote index won't be deleted.
+     */
+    public ElasticIndexCleaner(ElasticConnection elasticConnection, NodeStore nodeStore, int thresholdInSeconds) {
+        this.elasticConnection = elasticConnection;
+        this.nodeStore = nodeStore;
+        this.indexPrefix = elasticConnection.getIndexPrefix();
+        danglingRemoteIndicesMap = new HashMap<>();
+        this.threshold = thresholdInSeconds;
+    }
+
+    public void run() {
+        try {
+            NodeState root = nodeStore.getRoot();
+            InputStream inputStream;
+            inputStream = elasticConnection.getClient().getLowLevelClient()
+                    .performRequest(new Request("GET", "_cat/indices/" + elasticConnection.getIndexPrefix() + "*"))
+                    .getEntity().getContent();
+            List<String> remoteIndices = new BufferedReader(new InputStreamReader(inputStream))
+                    .lines().map(l -> l.split(" ")[2])
+                    .collect(Collectors.toList());
+            // remove entry of remote index names which don't exist now
+            List<String> externallyDeletedIndices = danglingRemoteIndicesMap.keySet().stream().filter(index -> !remoteIndices.contains(index)).collect(Collectors.toList());
+            externallyDeletedIndices.forEach(danglingRemoteIndicesMap::remove);
+            if (remoteIndices.isEmpty()) {
+                LOG.debug("No remote index found with prefix {}", indexPrefix);
+                return;
+            }
+            Set<String> existingIndices = new HashSet<>();
+            root.getChildNode(INDEX_DEFINITIONS_NAME).getChildNodeEntries().forEach(childNodeEntry -> {
+                PropertyState typeProperty = childNodeEntry.getNodeState().getProperty(IndexConstants.TYPE_PROPERTY_NAME);
+                if (typeProperty != null && typeProperty.getValue(Type.STRING).equals(ElasticIndexDefinition.TYPE_ELASTICSEARCH)) {
+                    String indexPath = "/" + INDEX_DEFINITIONS_NAME + "/" + childNodeEntry.getName();
+                    String remoteIndexName = ElasticIndexNameHelper.getRemoteIndexName(indexPrefix, childNodeEntry.getNodeState(), indexPath);
+                    if (remoteIndexName != null) {
+                        existingIndices.add(remoteIndexName);
+                    }
+                }
+            });
+            DeleteIndexRequest deleteIndexRequest = new DeleteIndexRequest();
+            for (String remoteIndexName : remoteIndices) {
+                if (!existingIndices.contains(remoteIndexName)) {
+                    Long curTime = System.currentTimeMillis();
+                    Long oldTime = danglingRemoteIndicesMap.putIfAbsent(remoteIndexName, curTime);
+                    if (threshold == 0 || (oldTime != null && curTime - oldTime >= TimeUnit.SECONDS.toMillis(threshold))) {
+                        deleteIndexRequest.indices(remoteIndexName);
```

Comment:
Calling `indices` multiple times does not add the indexes but replace the old one with the new one. You should be able to capture this with a unit test.
Instead, you should accumulate the indexes to remove in an array an create the `DeleteIndexRequest` with it.

Reply:
+1 to what Fabrizio wrote... indices method signature is indices(String... indices).
Actually the code you wrote will still work AFAIK... it's just that it will only remove one index at a time :-)

Reply:
right :)
I have corrected it now and modified test as well.